src/workerd/api/streams/writable.c++

```diff
@@ -365,6 +365,154 @@ private:
+
+  kj::Promise<void> write(const void* buffer, size_t size) override {
+    if (size == 0) return kj::READY_NOW;
+    return canceler.wrap(context.run(
+        [ref=weakRef->addRef(), buffer, size](Worker::Lock& lock) mutable
+            -> kj::Promise<void> {
+      jsg::Lock& js = lock;
+      auto& context = IoContext::current();
+      KJ_IF_SOME(obj, ref->tryGet()) {
+        auto source = jsg::BufferSource::wrap(js, const_cast<void*>(buffer), size,
+                                [](void*,size_t,void*) {}, nullptr);
+        v8::Local<v8::Value> value = source.getHandle(js);
+        return context.awaitJs(js, obj.getInner().write(js, value));
+      } else {
+        return cancellationException();
+      }
+    }));
+  }
+
+  kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) override {
+
+    auto amount = 0;
+    for (auto& piece : pieces) {
+      amount += piece.size();
+    }
+    if (amount == 0) return kj::READY_NOW;
+    return canceler.wrap(context.run(
+        [ref=weakRef->addRef(), pieces](Worker::Lock& lock) mutable
+            -> kj::Promise<void> {
+      jsg::Lock& js = lock;
+      auto& context = IoContext::current();
+      KJ_IF_SOME(obj, ref->tryGet()) {
+        // This is rather unfortunate because we need to write each piece individually,
+        // which is a separate JS promise :-( ... what we need here is the equivalent of
+        // Promise.all but there is no C++ API provided by v8 to implement that... so we
+        // need to be creative. Still, it's more efficient than individual writes since
+        // we do not need to drop the isolate lock between each piece.
+
+        kj::Maybe<jsg::Promise<void>> lastPromise = kj::none;
+        for (auto& piece : pieces) {
+          if (piece.size() == 0) continue;
+          auto source = jsg::BufferSource::wrap(js, const_cast<byte*>(piece.begin()), piece.size(),
+                                                [](void*,size_t,void*) {}, nullptr);
+          v8::Local<v8::Value> value = source.getHandle(js);
+          auto write = obj.getInner().write(js, value);
+          KJ_IF_SOME(promise, lastPromise) {
+            lastPromise = promise.then(js, [write=kj::mv(write)](jsg::Lock&) mutable {
+              return kj::mv(write);
+            });
+          } else {
+            lastPromise = kj::mv(write);
+          }
+        }
+        return context.awaitJs(js, kj::mv(KJ_ASSERT_NONNULL(lastPromise)));
+        return kj::READY_NOW;
+      } else {
+        return cancellationException();
+      }
+    }));
+  }
+
+  // TODO(perf): We can't properly implement tryPumpFrom(), which means that Cap'n Proto will
+  //   be unable to perform path shortening if the underlying stream turns out to be another capnp
+  //   stream. This isn't a huge deal, but might be nice to enable someday. It may require
+  //   significant refactoring of streams.
+
+  kj::Promise<void> whenWriteDisconnected() override {
+    return kj::NEVER_DONE;
```

Comment:
Is there any way to listen for the stream entering an error state?

Reply:
umm.. we may be able to attach to the `writer.closed` promise? Or maybe the `writer.ready` promise. Will have to experiment to see.

Reply:
I added some comments in the code about this